flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.flink.tests.util.flink.ClusterController;
+import org.apache.flink.tests.util.flink.FlinkResource;
+import org.apache.flink.tests.util.flink.FlinkResourceSetup;
+import org.apache.flink.tests.util.flink.LocalStandaloneFlinkResourceFactory;
+import org.apache.flink.util.FileUtils;
+import org.apache.flink.util.OperatingSystem;
+import org.apache.flink.util.TestLogger;
+import org.apache.flink.util.UserClassLoaderJarTestUtils;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.hdfs.MiniDFSCluster;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/** ITCase for adding remote jar. */
+@RunWith(Parameterized.class)
+public class AddRemoteJarITCase extends TestLogger {
+    private static final Logger LOG = LoggerFactory.getLogger(AddRemoteJarITCase.class);
+
+    private static final String ADD_REMOTE_JAR_E2E_SQL = "add_remote_jar_e2e.sql";
+
+    private static final String EXECUTE_SQL_RESULT =
+            "{\"before\":null,\"after\":{\"id\":1,\"content\":\"hello world\"},\"op\":\"c\"}";
+
+    public static final String GENERATED_LOWER_UDF_CLASS = "LowerUDF";
+
+    public static final String GENERATED_LOWER_UDF_CODE =
+            "public class "
+                    + "%s"
+                    + " extends org.apache.flink.table.functions.ScalarFunction {\n"
+                    + "  public String eval(String str) {\n"
+                    + "    return str.toLowerCase();\n"
+                    + "  }\n"
+                    + "}\n";
+
+    @ClassRule public static TemporaryFolder tempFolder = new TemporaryFolder();
+
+    private static final Path hadoopClasspath = TestUtils.getResource(".*hadoop.classpath");
+
+    @Parameterized.Parameters(name = "executionMode")
+    public static Collection<String> data() {
+        return Arrays.asList("streaming", "batch");
+    }
+
+    @Rule
+    public final FlinkResource flink =
+            new LocalStandaloneFlinkResourceFactory().create(FlinkResourceSetup.builder().build());
+
+    private final String executionMode;
+    private Path result;
+
+    private MiniDFSCluster hdfsCluster;
```

Comment:
I agree with you that a more common way is benefit. But the related tests with  HbaseResource and KafkaResource  are all ignored for that the tests maybe get stuck. If we download hdfs and setup a local hdfs cluster by command, I am afraid the process maybe get stuck someday. Since many ITCase use MiniDFSCluster, current implementation is a stable test. What do you think?

Reply:
Ok, I think your concern make sense to me.